### ISSUE TITLE
Incomplete type compilation error for hessian

### ISSUE THREAD

**Issue by mculbert** [stan-dev/math#2854]
(Not sure if this belongs in math, stan, or cmdstan.)

Compilation of the hessian method fails when the model includes a vectorized Normal distribution sampling statement for a given column of a matrix.

Minimal example:

```{r}
# Dev version of cmdstanr from github, with cmdstan 2.31.0
library(cmdstanr)
xi <- matrix(rnorm(20), 10, 2)
Y <- matrix(rnorm(200, xi[rep(1:10, 10),]), 100, 2)
stanData <- list(
  D = 2,
  W = 10,
  N = 100,
  i = rep(1:10, 10),
  Y = Y
)
model <- write_stan_file("
  data {
    int<lower=1> D;
    int<lower=1> W;
    int<lower=1> N;
    array[N] int<lower=1, upper=W> i;
    matrix[N,D] Y;
  }
  parameters {
    matrix[W,D] xi;
  }
  model {
    for (d in 1:D) Y[,d] ~ normal(xi[i,d], 1);
    for (d in 1:D) xi[,d] ~ normal(0, 1);
  }
")
stanmodel <- cmdstan_model(model, compile_hessian_method=T)
fit <- stanmodel$optimize(data=stanData)
fit$init_model_methods(hessian=T)
```

Yields the following compilation error:

```
In file included from /home/mculbert/.cmdstan/cmdstan-2.31.0/stan/lib/stan_math/stan/math/fwd/functor.hpp:8,
                 from /home/mculbert/.cmdstan/cmdstan-2.31.0/stan/lib/stan_math/stan/math/mix.hpp:11,
                 from /home/mculbert/.cmdstan/cmdstan-2.31.0/stan/src/stan/model/hessian.hpp:4,
                 from file3890970bff175.cpp:581:
/home/mculbert/.cmdstan/cmdstan-2.31.0/stan/lib/stan_math/stan/math/fwd/functor/operands_and_partials.hpp: In instantiation of ‘class stan::math::operands_and_partials<const Eigen::Block<Eigen::Matrix<stan::math::fvar<stan::math::var_value<double> >, -1, -1>, -1, 1, true>&, const int&, const int&, double, double, stan::math::fvar<stan::math::var_value<double> > >’:
/home/mculbert/.cmdstan/cmdstan-2.31.0/stan/lib/stan_math/stan/math/prim/prob/normal_lpdf.hpp:73:57:   required from ‘stan::return_type_t<T, L, U> stan::math::normal_lpdf(const T_y&, const T_loc&, const T_scale&) [with bool propto = true; T_y = Eigen::Block<Eigen::Matrix<stan::math::fvar<stan::math::var_value<double> >, -1, -1>, -1, 1, true>; T_loc = int; T_scale = int; stan::require_all_not_nonscalar_prim_or_rev_kernel_expression_t<T_y, T_scale_succ, T_scale_fail>* <anonymous> = 0; stan::return_type_t<T, L, U> = stan::math::fvar<stan::math::var_value<double> >]’
file3890970bff175.cpp:192:46:   required from ‘stan::scalar_type_t<T2> model_402b3d620ae5dcf64fb257245fdd0a41_model_namespace::model_402b3d620ae5dcf64fb257245fdd0a41_model::log_prob_impl(VecR&, VecI&, std::ostream*) const [with bool propto__ = true; bool jacobian__ = true; VecR = Eigen::Matrix<stan::math::fvar<stan::math::var_value<double> >, -1, 1>; VecI = Eigen::Matrix<int, -1, 1>; stan::require_vector_like_t<VecR>* <anonymous> = 0; stan::require_vector_like_vt<std::is_integral, VecI>* <anonymous> = 0; stan::scalar_type_t<T2> = stan::math::fvar<stan::math::var_value<double> >; std::ostream = std::basic_ostream<char>]’
file3890970bff175.cpp:401:77:   required from ‘T_ model_402b3d620ae5dcf64fb257245fdd0a41_model_namespace::model_402b3d620ae5dcf64fb257245fdd0a41_model::log_prob(Eigen::Matrix<T_job_param, -1, 1>&, std::ostream*) const [with bool propto__ = true; bool jacobian__ = true; T_ = stan::math::fvar<stan::math::var_value<double> >; std::ostream = std::basic_ostream<char>]’
file3890970bff175.cpp:598:53:   required from ‘T hessian_wrapper<M>::operator()(const Eigen::Matrix<T, -1, 1>&) const [with T = stan::math::fvar<stan::math::var_value<double> >; M = model_402b3d620ae5dcf64fb257245fdd0a41_model_namespace::model_402b3d620ae5dcf64fb257245fdd0a41_model&]’
/home/mculbert/.cmdstan/cmdstan-2.31.0/stan/lib/stan_math/stan/math/mix/functor/hessian.hpp:61:26:   required from ‘void stan::math::hessian(const F&, const Eigen::Matrix<double, -1, 1>&, double&, Eigen::Matrix<double, -1, 1>&, Eigen::Matrix<double, -1, -1>&) [with F = hessian_wrapper<model_402b3d620ae5dcf64fb257245fdd0a41_model_namespace::model_402b3d620ae5dcf64fb257245fdd0a41_model&>]’
file3890970bff175.cpp:616:53:   required from here
/home/mculbert/.cmdstan/cmdstan-2.31.0/stan/lib/stan_math/stan/math/fwd/functor/operands_and_partials.hpp:73:54: error: ‘stan::math::operands_and_partials<Op1, Op2, Op3, Op4, Op5, stan::math::fvar<Dx> >::edge1_’ has incomplete type
   73 |   internal::ops_partials_edge<Dx, std::decay_t<Op1>> edge1_;
      |                                                      ^~~~~~
In file included from /home/mculbert/.cmdstan/cmdstan-2.31.0/stan/lib/stan_math/stan/math/prim/fun/binomial_coefficient_log.hpp:13,
                 from /home/mculbert/.cmdstan/cmdstan-2.31.0/stan/lib/stan_math/stan/math/prim/fun.hpp:33,
                 from /home/mculbert/.cmdstan/cmdstan-2.31.0/stan/lib/stan_math/stan/math/rev/fun/multiply.hpp:7,
                 from /home/mculbert/.cmdstan/cmdstan-2.31.0/stan/lib/stan_math/stan/math/rev/fun/elt_multiply.hpp:9,
                 from /home/mculbert/.cmdstan/cmdstan-2.31.0/stan/lib/stan_math/stan/math/rev/fun.hpp:55,
                 from /home/mculbert/.cmdstan/cmdstan-2.31.0/stan/lib/stan_math/stan/math/rev.hpp:10,
                 from /home/mculbert/.cmdstan/cmdstan-2.31.0/stan/lib/stan_math/stan/math.hpp:19,
                 from /home/mculbert/.cmdstan/cmdstan-2.31.0/stan/src/stan/model/model_header.hpp:4,
                 from file3890970bff175.cpp:3:
/home/mculbert/.cmdstan/cmdstan-2.31.0/stan/lib/stan_math/stan/math/prim/functor/operands_and_partials.hpp:21:7: note: declaration of ‘class stan::math::internal::ops_partials_edge<stan::math::var_value<double>, Eigen::Block<Eigen::Matrix<stan::math::fvar<stan::math::var_value<double> >, -1, -1>, -1, 1, true>, void>’
   21 | class ops_partials_edge;
      |       ^~~~~~~~~~~~~~~~~
make: *** [/usr/lib/R/etc/Makeconf:178: file3890970bff175.o] Error 1
g++ -std=gnu++14 -I"/usr/share/R/include" -DNDEBUG   -I"/home/mculbert/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include" -I"/home/mculbert/R/x86_64-pc-linux-gnu-library/4.2/RcppEigen/include" -I"/tmp/RtmpA07Nup/sourceCpp-x86_64-pc-linux-gnu-1.0.9"   -std=c++1y -pthread -D_REENTRANT -Wno-sign-compare -Wno-ignored-attributes -I'/home/mculbert/.cmdstan/cmdstan-2.31.0/stan/lib/stan_math/lib/tbb_2020.3/include' -O3 -I'/home/mculbert/.cmdstan/cmdstan-2.31.0/src' -I'/home/mculbert/.cmdstan/cmdstan-2.31.0/stan/src' -I'/home/mculbert/.cmdstan/cmdstan-2.31.0/stan/lib/rapidjson_1.1.0/' -I'/home/mculbert/.cmdstan/cmdstan-2.31.0/lib/CLI11-1.9.1/' -I'/home/mculbert/.cmdstan/cmdstan-2.31.0/stan/lib/stan_math/' -I'/home/mculbert/.cmdstan/cmdstan-2.31.0/stan/lib/stan_math/lib/eigen_3.3.9' -I'/home/mculbert/.cmdstan/cmdstan-2.31.0/stan/lib/stan_math/lib/boost_1.78.0' -I'/home/mculbert/.cmdstan/cmdstan-2.31.0/stan/lib/stan_math/lib/sundials_6.1.1/include' -I'/home/mculbert/.cmdstan/cmdstan-2.31.0/stan/lib/stan_math/lib/sundials_6.1.1/src/sundials' -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-a3XuZ5/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c file3890970bff175.cpp -o file3890970bff175.o
Error in Rcpp::sourceCpp(code = code, env = env, verbose = verbose) : 
  Error 1 occurred building shared library.
```

Error seems to be for the `xi ~ normal` statement, but not the `Y ~ normal` statement.

Compilation was fine if `xi` is declared as a vector, instead of a matrix.

**Comment by andrjohns** [comment]
Thanks for reporting (and yep this is the right place)! I'll have a look into this

**Comment by WardBrian** [comment]
Hm, the issue seems to go away with what would be (I believe) a semantically equivalent model:

```stan
data {
  int<lower=1> D;
  int<lower=1> W;
  int<lower=1> N;
  array[N] int<lower=1, upper=W> i;
  matrix[N,D] Y;
}
parameters {
  matrix[W,D] xi;
}
model {
  for (d in 1:D) to_vector(Y[,d]) ~ normal(xi[i,d], 1);
  for (d in 1:D) to_vector(xi[,d]) ~ normal(0, 1);
}
```

So it seems the issue may be specific to `Eigen::Block` expression types

**Comment by andrjohns** [comment]
Closed by #2917
